test/performance/io/format_fastq_benchmark.cpp

```diff
@@ -0,0 +1,285 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+// @author: Clemens Cords <clemens.cords@fu-berlin.de>
+
+#include <benchmark/benchmark.h>
+#include <cstring>
+#include <fstream>
+
+#define SEQAN2_HAS_SEQAN3 __has_include(<seqan/seq_io.h>)
+#if SEQAN2_HAS_SEQAN3
+    #include <seqan/seq_io.h>
+#endif
+
+#include <seqan3/alphabet/nucleotide/dna5.hpp>
+#include <seqan3/alphabet/quality/phred42.hpp>
+#include <seqan3/io/sequence_file/format_fastq.hpp>
+#include <seqan3/io/sequence_file/input.hpp>
+#include <seqan3/range/view/to_char.hpp>
+#include <seqan3/test/performance/sequence_generator.hpp>
+#include <seqan3/std/filesystem>
+
+using namespace seqan3;
+
+unsigned int const SEED = 1234;
+
+// ============================================================================
+// generate fastq file
+// ============================================================================
+constexpr std::size_t SEQUENCE_LENGTH = 50; //length of nucleotide and quality sequence
+constexpr std::size_t N_ENTRIES_IN_FILE = 4069; // default number of 3-line entries
+
+// if a file of length m is requested once, that file is stored so
+// any functions later on that needs the exact same file can access it
+std::map<std::size_t, std::string> file_dict;
+
+std::string get_file(std::size_t n_entries)
+{
+    bool has_been_init = file_dict.find(n_entries) != file_dict.end();
+
+    if (!has_been_init)
+    {
+        std::cout << std::endl; // sic, helps formatting print output
```

Comment:
I don't understand this.

Reply:
this creates a newline in between benchmark calls with different template parameters like so:
```
fastq_write_to_stream                     16099 ns        16098 ns        44227

fastq_read_from_stream<1000>              90514 ns        90516 ns         7443
fastq_read_from_disk<1000>             84545863 ns     84548582 ns            8
fastq_read_from_stream_seqan2<1000>       15878 ns        15879 ns        43251
fastq_read_from_disk_seqan2<1000>       6828959 ns      6829089 ns           98

fastq_read_from_stream<5000>             124365 ns       124369 ns         5683
fastq_read_from_disk<5000>            387559999 ns    387567259 ns            2
fastq_read_from_stream_seqan2<5000>       51202 ns        51203 ns        13720
fastq_read_from_disk_seqan2<5000>      34893085 ns     34894137 ns           20
```
It is obviously not needed but I find it makes the output nicer to read and compare and does not impact performance or functionality in any way

Reply:
Please don't use that.

Furthermore, what happens if you change the output format `--benchmark_format=json`?

Reply:
I really would like spacing when printing to the command line actually. Is there another way of doing this?

Reply:
I don't know. For me it sounds like a feature request.

Reply:
I don't think there isn't another way, to guarantee that the newline is printed in the right place one would need a function like:
```cpp
void newline()
{
    // print newline
    return;
}
#define NEWLINE BENCHMARK(newline);

BENCHMARK(A1)
BENCHMARK(A2)
NEWLINE
BENCHMARK(B1)
BENCHMARK_MAIN()
```
But this isn't possible since the ``BENCHMARK`` macro will only accepts functions with the signature ``void (benchmark::state&)`` afaik. So there isn't a way to do this without modifying google benchmark I think.

Reply:
Even this wouldn't work because you can't control how often ``benchmark(...)`` would be invoked

```cpp
// some header
void newline()
{
    // print newline
    return;
}
#define NEWLINE newline();

// benchmark.cpp
void benchmark(benchmark::State & state) 
{
    (...)
    for (auto _ : state)
        (...)
        
    NEWLINE;
}

BENCHMARK(benchmark);
BENCHMARK(other_benchmark_function);
BENCHMARK_MAIN();
```